Docs Update -- User Guide Sections by MS

Shoeboxam opened the pull request:
Closes #269.

- fixed python doctests
- adjusted makefile for testing python doctests manually with `make doctest-python`
- added python doctests to the smoke-test CI
- added documentation files:
    * core-structures.rst
    * supporting-elements.rst
    * combinators.rst
    * transformations.rst
    * measurements.rst
    * application-structure.rst

andrewvyrros commented on docs/source/user/supporting-elements.rst:
```diff
@@ -1,20 +1,146 @@
+The typical perturbation is the adjustment made to a dataset by adding or removing a user.
+
+
+Distances
```
I'd rename this section something like, Putting It Together. Also, a little more explanation of why we're doing all this would help cement the concepts.

(Also have a section on Distances, as a concept, above after Metrics and Measures.)

andrewvyrros commented on docs/source/user/supporting-elements.rst:
```diff
@@ -1,20 +1,146 @@
+If we know the class has three exams, then we know each student may contribute at most three records, so `d_in` should be 3.
+
+We can use this information in a measurement that casts, imputes, clamps, resizes, averages, and then adds noise.
+We can use the relation on this chained measurement to conduct a binary search to find `d_out`, where `d_in` = 3.
+
+It is also possible to use `binary_search_chain` to find a noise scale parameter that satisfies a given budget.
```
Hoo boy, the reason for binary_search is gonna be really hard for people to understand! It's tough to explain, but I think we need some more verbiage about what's going on and why you need to do it.

andrewvyrros commented:
General style question: Should we use code styling (rst double-backquotes) for symbols like `Measurement` and `make_base_laplace`? And/or, should we link to the API docs (maybe just on the first usage)?

andrewvyrros commented on docs/source/user/core-structures.rst:
```diff
@@ -1,28 +1,98 @@
+Each Measurement consists of a privacy relation and a function.
+
+The privacy relation is used to check if the function is `d_out`-DP on `d_in`-close inputs.
+The privacy relation is qualified by an input metric and an output measure, which give units to the distances `d_in` and `d_out`.
```
Also, where we're referencing concepts we haven't encountered yet (especially for domains and relations which don't appear until the next doc), it'd be good to link to the section that covers them.

andrewvyrros commented on docs/source/user/core-structures.rst:
```diff
@@ -1,28 +1,98 @@
 -----------
 
 A Measurement is a randomized mapping from datasets to outputs of an arbitrary type.
+Each Measurement consists of a privacy relation and a function.
```
Should we use code styling (rst double-backquotes) for symbols like Measurement? And/or, should we link to the API docs (maybe just on the first usage)?

andrewvyrros commented:
General suggestion: Where we have to mention a topic that isn't explained until later, it'd be good to link to the future section. For instance, when in Measurements, first time Domain is mentioned, link to the section on Domains.

andrewvyrros commented:
Is `test_docs.py` (empty) meant to be here? (I added a `make doctest` target that uses `sphinx doctest`, was planning to incorporate that into #253, but maybe there's a better way to handle that?)

andrewvyrros reviewed (commented):
Thanks for doing this! So many concepts to document... This is a great start, will really help people get up to speed on the library.

Most of my suggestions are around providing more general explanations of things before getting to specifics. It think this will help people who, unlike us, have never encountered these concepts before.

Shoeboxam commented:
I updated this with:
- fixed python doctests
- adjusted makefile for testing python doctests manually with `make doctest-python`
- rewrote existing sections based on Andy's feedback
- added combinators.rst, transformations.rst, measurements.rst and application-structure.rst

andrewvyrros commented on docs/source/user/transformations.rst:
```diff
@@ -0,0 +1,246 @@
+If we were to allow aggregations to propagate null,
+then aggregations provide a non-differentially-private bit revealing the existence of nullity in the dataset.
+If we were to implicitly drop nulls from sized aggregations, then the sensitivity of non-null individuals is underestimated.
+Therefore, aggregators must fed completely non-null data.
```
```suggestion
Therefore, aggregators must be fed completely non-null data.
```

andrewvyrros commented on docs/source/user/transformations.rst:
```diff
@@ -0,0 +1,246 @@
+This documentation gives a high-level overview of the transformations that are available in the library.
+Refer to the :ref:`transformation` section for an explanation of what a transformation is.
+
+As covered in the :ref:`chaining` section, the inner domains need to match when chaining.
```
```suggestion
As covered in the :ref:`chaining` section, the intermediate domains need to match when chaining.
```

andrewvyrros commented on docs/source/user/measurements.rst:
```diff
@@ -0,0 +1,79 @@
+As covered in the :ref:`chaining` section, the inner domains and metrics need to match when chaining.
+This means you will need to choose a measurement that chains with your :ref:`aggregator <aggregators>`.
+
+In the following table, the scalar-valued and vector-valued versions of each measurement are listed separately.
```
Maybe add a little more detail explaining that choice of `D` determines which you get.

andrewvyrros commented on docs/source/user/application-structure.rst:
```diff
@@ -3,3 +3,143 @@ Application Structure
+
+.. _accuracy:
+
+Accuracy
```
Maybe call this Determining Accuracy?

Also, I think I'd move this section after the binary search section.

andrewvyrros commented on docs/source/user/application-structure.rst:
```diff
@@ -3,3 +3,143 @@ Application Structure
+
+.. _binary-search:
+
+Binary Search
```
I'd call this Choosing Parameters or Choosing Noise Scale.

Also, I think this section should come first, as it's more fundamental than accuracy.

Shoeboxam replied:
I don't like "choosing noise scale" because it's really "choosing any free parameter". How about "Parameter Search"?

andrewvyrros commented on docs/source/user/limitations.rst:
```diff
@@ -1,3 +1,5 @@
+.. _limitations:
```
Hmmm that's weird, I thought I had automatic heading links working, did that break?

Shoeboxam replied:
I tried adding it once and autoref complained about duplicate section names all throughout the documentation. Interested to hear how you added it. But no, I can't remove it on the current configuration in my branch.

andrewvyrros commented on .github/workflows/smoke-test.yml:
```diff
@@ -97,3 +97,10 @@ jobs:
+        run: |
+          cd ../docs
+          python -m pip install -r requirements.txt
+          make html
```
Make will do this for you!
```suggestion
```

andrewvyrros commented on docs/Makefile:
```diff
@@ -35,11 +35,17 @@ html:
+doctest-rust:
+	cd ../rust && $(CARGO) test --doc
+
+doctest-python:
```
Dependencies are your friend!
```suggestion
doctest-python: html
```

andrewvyrros commented on docs/Makefile:
```diff
@@ -35,11 +35,17 @@ html:
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
 
 doctest:
```
Dependencies are your friend!
```suggestion
doctest: doctest-rust doctest-python
```

andrewvyrros commented on docs/Makefile:
```diff
@@ -35,11 +35,17 @@ html:
 doctest:
-	cd ../rust && $(CARGO) test --doc
-	$(SPHINXBUILD) $(SPHINXOPTS) -b doctest source $(BUILDDIR)/doctest
+	make doctest-rust
+	make doctest-python
```
```suggestion
```

andrewvyrros reviewed (changes requested):
Wow! This is PHENOMENAL! Fantastic stuff, really will help users understand the library.

There are a few smallish things that I think are worth addressing, but overall this is great.

Shoeboxam commented on docs/source/user/core-structures.rst:
```diff
@@ -7,22 +9,115 @@ Core Structures
+
+A unifying perspective towards OpenDP is that OpenDP is a system for relating:
+
+#. an upper bound on distance between neighboring function inputs (to)
```
it just makes the text flow when you read it out loud

andrewvyrros reviewed (approved):
Outstanding!
